File: the-nest-backend/app/models/notification.py
import uuid
from sqlalchemy import Boolean, Column, DateTime, Enum, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.sql import func

from app.database import Base

NotificationType = Enum(
    "like",
    "repost",
    "reply",
    "follow",
    "mention",
    name="notificationtype",
)


class Notification(Base):
    __tablename__ = "notifications"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    recipient_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    actor_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False)
    type = Column(NotificationType, nullable=False)
    post_id = Column(UUID(as_uuid=True), ForeignKey("posts.id", ondelete="CASCADE"), nullable=True)
    is_read = Column(Boolean, default=False, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now(), nullable=False)

File: the-nest-frontend/types/notification.ts
export interface NotificationActor {
  username: string
  display_name: string
  avatar_url: string | null
}

export interface Notification {
  id: string
  type: 'like' | 'repost' | 'reply' | 'follow' | 'mention'
  is_read: boolean
  created_at: string
  post_id: string | null
  actor: NotificationActor
}
